### PR TITLE
Add createdAt to createAccount RPC

### PR DESCRIPTION
**Pull request by danield9tqh**
## Summary
Allow account creation to take a createdAt object

## Testing Plan
Unit tests

## Documentation

Does this change require any updates to the Iron Fish Docs (ex. [the RPC API
Reference](https://ironfish.network/docs/onboarding/rpc/chain))? If yes, link a
related documentation pull request for the website.

```
[ ] Yes
```

## Breaking Change

Is this a breaking change? If yes, add notes below on why this is breaking and label it with `breaking-change-rpc` or `breaking-change-sdk`.

```
[ ] Yes
```

### CODE REVIEW

File: ironfish/src/rpc/routes/wallet/createAccount.ts [hunk 1]
```diff
@@ -45,10 +59,23 @@ routes.register<typeof CreateAccountRequestSchema, CreateAccountResponse>(
   CreateAccountRequestSchema,
   async (request, context): Promise<void> => {
     AssertHasRpcContext(request, context, 'wallet')
+    if (
+      request.data.createdAt?.hash === undefined ||
+      request.data.createdAt?.sequence === undefined
+    ) {
+      request.data.createdAt = undefined
```
Comment (hughy) [inline ironfish/src/rpc/routes/wallet/createAccount.ts:66]: will yup validation enforce both fields being defined this if we add `defined()` to the `hash` and `sequence` fields?

File: ironfish/src/rpc/routes/wallet/createAccount.ts [hunk 2]
```diff
@@ -18,7 +18,15 @@ import { AssertHasRpcContext } from '../rpcContext'
...
  */
 
-export type CreateAccountRequest = { name: string; default?: boolean }
+export type CreateAccountRequest = {
+  name: string
+  default?: boolean
+  createdAt?: {
+    hash: string
+    sequence: number
+  }
```
Comment (hughy) [inline ironfish/src/rpc/routes/wallet/createAccount.ts:27]: we're in the process of deprecating `hash` as part of `createdAt`. the `importAccount` RPC only uses a number for `createdAt`. should we do the same here?

File: ironfish/src/rpc/routes/wallet/createAccount.ts [hunk 3]
```diff
@@ -45,10 +59,23 @@ routes.register<typeof CreateAccountRequestSchema, CreateAccountResponse>(
   CreateAccountRequestSchema,
   async (request, context): Promise<void> => {
     AssertHasRpcContext(request, context, 'wallet')
+    if (
+      request.data.createdAt?.hash === undefined ||
+      request.data.createdAt?.sequence === undefined
+    ) {
+      request.data.createdAt = undefined
+    }
```
Comment (NullSoldier) [inline ironfish/src/rpc/routes/wallet/createAccount.ts:67]: You can delete this entire block of code according to your spec above.

File: ironfish/src/rpc/routes/wallet/createAccount.ts [hunk 4]
```diff
@@ -45,10 +59,23 @@ routes.register<typeof CreateAccountRequestSchema, CreateAccountResponse>(
...
+
+    const createdAt = request.data.createdAt && {
+      hash: Buffer.from(request.data.createdAt.hash, 'hex'),
+      sequence: request.data.createdAt.sequence,
+    }
+
+    const setDefault = !context.wallet.hasDefaultAccount || (request.data.default ?? false)
```
Comment (NullSoldier) [inline ironfish/src/rpc/routes/wallet/createAccount.ts:74]: random nit: but I prefer small refactors like this split out to other PRs. This is not related to the PR.

Comment (NullSoldier) [reply ironfish/src/rpc/routes/wallet/createAccount.ts:74]: Also because now that this code changed, the code is actually very strange that the RPC is auto setting the default account if one is not set when an account is created. This should absolutely not be happening and should be removed in another PR.

File: ironfish/src/rpc/routes/wallet/createAccount.ts (hunk 3 above)
Comment (danield9tqh) [reply ironfish/src/rpc/routes/wallet/createAccount.ts:67]: The yup spec? Yup doesn't allow required fields nested in an optional object. So if we passed in the following
```
{ "name": "account1" }
```
the request object `data` would come in as
```
{
  "name": "account1"
  "createdAt": {
    hash: undefined
    sequence: undefined
  }
}
```

I could probably make this code a little cleaner but it would be more complex if we actually wanted to validate that if you pass in `createdAt` field that both `hash` and `sequence` are defined. This is one of the main problems with yup I've run into. It gets even more difficult if you want to make `createdAt` nullable (which it techincally should be since you can set `createdAt` to null in the wallet)

File: ironfish/src/rpc/routes/wallet/createAccount.ts (hunk 4 above)
Comment (danield9tqh) [reply ironfish/src/rpc/routes/wallet/createAccount.ts:74]: I can move this refactor to another PR. I'm still confused on how it changes the functionality though. In the previous code if there was no default account before this RPC call, it would set this account as the default. Seems the same as what this code is doing

Comment (NullSoldier) [reply ironfish/src/rpc/routes/wallet/createAccount.ts:74]: There's no change in functionality here. I re-read previous comments and don't see any mention of that. That's the issue I'm mentioning, it's a refactor that doesn't change functionality for an orthagonal piece of code unrelated to this change. (using the internal default setting rather than duplicating it)

File: ironfish/src/rpc/routes/wallet/create.test.slow.ts [hunk 5]
```diff
@@ -88,4 +93,34 @@ describe('Route wallet/create', () => {
...
+    expect(account).toMatchObject({
+      name: name,
+      publicAddress: response.content.publicAddress,
+      createdAt: {
+        hash: Buffer.from(createdAt.hash, 'hex'),
+        sequence: createdAt.sequence,
+      },
+    })
+  })
 })
```
Comment (NullSoldier) [inline ironfish/src/rpc/routes/wallet/create.test.slow.ts:126]: This file was for testing /create but someone changed it to use the new RPC route which means the old /create isn't being tested. On top of that, the file name is now wrong.

File: ironfish/src/rpc/routes/wallet/createAccount.ts (hunk 3 above)
Comment (NullSoldier) [reply ironfish/src/rpc/routes/wallet/createAccount.ts:67]: I pulled the PR down and it works just fine for me, you just have the YUP schema defined incorrectly. For optional sub fields you need this syntax.

```
export const CreateAccountRequestSchema: yup.ObjectSchema<CreateAccountRequest> = yup
  .object({
    name: yup.string().defined(),
    default: yup.boolean().optional(),
    createdAt: yup
      .object({
        hash: yup.string().defined(),
        sequence: yup.number().defined(),
      })
      .notRequired()
      .default(undefined),
  })
  .defined()
```

Once you do this, you can delete the entire block of code setting the createdAt as undefined.

File: ironfish/src/rpc/routes/wallet/createAccount.ts (hunk 2 above)
Comment (danield9tqh) [reply ironfish/src/rpc/routes/wallet/createAccount.ts:27]: So shouldn't make a difference if we just set it to an empty buffer? `Buffer.alloc(32, 0)`

File: ironfish/src/rpc/routes/wallet/create.test.slow.ts (hunk 5 above)
Comment (NullSoldier) [reply ironfish/src/rpc/routes/wallet/create.test.slow.ts:146]: Ok we're going to fix this in another PR.

**Review by NullSoldier — APPROVED** [review]

**Review by hughy — APPROVED** [review]
